### PR TITLE
feat(exp/query): Add support for disk based LRU cache 

### PR DESCRIPTION
**Pull request by kavirajk**
### What's being changed:
Changes
1. Move LSM fetching to different component called `LSMFetcher`
2. LSMFetcher can work with or without cache.
3. Introduce simple LRU based local disk cache
4. Integerate everything with query API
5. Remove old EnsureLSM and metadata subscription related code.

#### API details

LSMFetcher owns bringing LSM data of tenant either from cache or upstream store. It has two API
1. `NewLSMFetcher()`
2. `NewLSMFetcherWithCache()`

If you use without cache, every single tenant requests -> makes call to upstream store (s3)

Diskcache takes care of caching based on version. It has following APIs
1. `NewDisCache()`
2. `Tenant()` -> Get tenant data
3. `AddTenant()` -> Put tenant data
4. `BasePath()` -> basepath of cache directory.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: exp/query/config.go [hunk 1]
```diff
@@ -22,16 +22,15 @@ type Config struct {
...
-	DataPath string `long:"datapath" description:"place to look for tenant data after downloading it from object storage" default:"/tmp"`
+	DataPath string `long:"datapath" description:"place to look for tenant data after downloading it from object storage" default:"/tmp/weaviate"`
 
 	VectorizerAddr string `long:"vectorize-addr" description:"vectorizer address to be used to vectorize near-text query" default:"0.0.0.0:9999"`
 
 	// MetadataGRPCAddress is the host which will be used to connect to the metadata node's gRPC server.
 	// Note that this should be replaced later to avoid having a single metadata node as a single point of failure.
 	// If MetadataGRPCAddress is empty, the querier will connect to localhost.
 	MetadataGRPCAddress string `long:"metadata.grpc.address" description:"metadata grpc address" default:":9050"`
-	// AlwaysFetchObjectStore flag ignore what version does local querier has and fetch from
```
Comment (kavirajk) [inline exp/query/config.go:33]: renamed `AlwaysFetchObjectStore` -> `NoCache`. Because made sense.

File: exp/query/lsm_fetcher.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	lsmPath := path.Join(tenantPath, defaultLSMRoot)
+
+	store, err := lsmkv.New(lsmPath, lsmPath, l.log, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
+	}
+
+	return store, lsmPath, nil
```
Comment (nathanwilk7) [inline exp/query/lsm_fetcher.go:130]: Shouldn't this method return early if there is a cache hit?

The current implementation pushes a new entry onto the evictlist for every search (unbounded) and also calls `usage()` on every search.

Can you do something like?

```suggestion
	var (
		basePath          = l.basePath
		tenantPath string = ""
	)

	if l.cache != nil {
		basePath = l.cache.BasePath()
		c, err := l.cache.Tenant(collection, tenant)
		if err == nil && version != 0 && c.Version >= version {
			tenantPath = c.AbsolutePath()
			// return lsmkv.New here
	}

	tenantPath = path.Join(
		basePath,
		strings.ToLower(collection),
		strings.ToLower(tenant),
	)
	// src - s3://<collection>/<tenant>/<node>/
	// dst (local) - <data-path/<collection>/<tenant>
	l.log.WithFields(logrus.Fields{
		"collection": collection,
		"tenant":     tenant,
		"nodeName":   nodeName,
		"dst":        tenantPath,
	}).Debug("starting download to path")
	if err := l.upstream.DownloadToPath(ctx, collection, tenant, nodeName, tenantPath); err != nil {
		return nil, "", err
	}

	// populate the cache before returning
	if l.cache != nil {
		if err := l.cache.AddTenant(collection, tenant, version); err != nil {
			return nil, "", fmt.Errorf("failed populate the cache: %w", err)
		}
	}

	lsmPath := path.Join(tenantPath, defaultLSMRoot)

	store, err := lsmkv.New(lsmPath, lsmPath, l.log, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
	if err != nil {
		return nil, "", fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
	}

	return store, lsmPath, nil
```

Comment (nathanwilk7) [reply exp/query/lsm_fetcher.go:130]: A related issue we should leave a future TODO for is that the current implementation will download the same tenant for each versioned search request until the tenant version is downloaded and in the cache (eg if 10 requests come for the same tenant come at ~the same time, this will run 10 downloads from s3).

File: exp/query/cache.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	_, err := os.Stat(tc.AbsolutePath())
+	if errors.Is(err, os.ErrNotExist) {
+		return errors.Join(ErrTenantDirectoryFound, err)
+	}
+
+	c := d.evictList.PushFront(tc)
+	d.tenants[d.TenantKey(tc.Collection, tc.TenantID)] = c
+
+	// evit if the size is getting filled up
+	// TODO(kavi): Worth doing it on different go routine? But will get complex with correctness
```
Comment (nathanwilk7) [inline exp/query/cache.go:94]: i'm fine with leaving this usage call here for now, but as a follow up can you add some metrics for how often it's called/how long it takes?

File: exp/query/cache.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	if d.usage() > d.maxCap {
+		if err := d.evict(); err != nil {
```
Comment (nathanwilk7) [inline exp/query/cache.go:96]: What happens if we evict a tenant in parallel while that tenant is being searched or downloaded to? I think we can come back to this later, but should leave a TODO in code or create an issue to handle it

File: exp/query/lsm_fetcher.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		)
+		// src - s3://<collection>/<tenant>/<node>/
+		// dst (local) - <data-path/<collection>/<tenant>
+		l.log.WithFields(logrus.Fields{
+			"collection": collection,
+			"tenant":     tenant,
+			"nodeName":   nodeName,
+			"dst":        tenantPath,
+		}).Debug("starting download to path")
+		if err := l.upstream.DownloadToPath(ctx, collection, tenant, nodeName, tenantPath); err != nil {
```
Comment (nathanwilk7) [inline exp/query/lsm_fetcher.go:110]: If we're downloading a new version of this tenant, don't we need to remove the old one? Otherwise we would have all the files merged for each version each time we download?

**Review by nathanwilk7 — CHANGES_REQUESTED** [review]
Thanks for doing this, I really like the fetcher/cache abstractions. I think this is ready to merge once the cache hit->evictlist/usage issue has been resolved and we've discussed my concerns with downloading multiple versions?

File: exp/query/lsm_fetcher.go (hunk 5 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go:110]: Good point. I would do following instead of separate step to remove the old tenant.

Say we have some version under `/cache/collection/tenant-1/` and we found more latest version. All downloads should happen in some tmp location (`/tmp/collection/tenant-1/`) first and atomically replace old with new (`mv /tmp/collection/tenant-1/` `cache/collection/tenant-1/`) using Go's `os.Rename` (which is atomic as long as both source and destination in same filesystem)

That's usually recommended way to do such things.

Rationale being, I don't want to maintain multiple versions of `/collection/tenant-1/` in local cache. That way there will always be one copy of latest tenant data at any point in time. Easy to mental model IMHO.

wdyt?

File: exp/query/lsm_fetcher.go (hunk 2 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go:130]: Hmm. another good question 👍 

I actually have an idea to handle this case using Go's [singleflight](https://pkg.go.dev/golang.org/x/sync/singleflight) package. 

tldr; we create single `singleflight.Group` for each `key` of the cache. So that if 10 concurrent requests comes in for single key and if it's a cache miss, then instead of spanning out 10 requests to upstream, we do one upstream call and share the results with remaining requests (other 9 would wait meanwhile).

It's very handy package I have used it in the past. I didn't do it straightway to to keep the PR simple for now. Happy to add a TODO. good catch. Thanks.

File: exp/query/cache.go (hunk 4 above)
Comment (kavirajk) [reply exp/query/cache.go:96]: > if we evict a tenant in parallel while that tenant is being searched or downloaded to?
This case happens, only say a particular tenant is not being used for long time (till eviction kicked in) and this tenant reached the end of eviction list and suddenly the request for this tenant comes in. What happens here is, if eviction wins the race, we will do one extra upstream call.

This is the inherent problem with LRU eviction policy. I don't think any sane solution for it unless I miss something.

File: exp/query/lsm_fetcher.go (hunk 2 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go:130]: Also worth mentioning that this solves the multiple in-flight requests problem only in single process (say requests coming to single querier replica in our case). 

If say we have 10 in-flight requests for same tenant but being handled by multiple queriers, still each querier have to make at most one upstream call.

There are ways to avoid this as well (say using like [Groupcache)](https://github.com/golang/groupcache) but adds more complexity and definitely not worth doing unless we have some real usage data IMHO

Comment (reyreaud-l) [reply exp/query/lsm_fetcher.go:130]: > If say we have 10 in-flight requests for same tenant but being handled by multiple queriers, still each querier have to make at most one upstream call.

Agree, it's ok if request lands on different queriers that queriers will all hit S3. I like the singleflight pkg I had never seen it before.

File: exp/query/lsm_fetcher.go (hunk 5 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go]: I actually ended up using `/<collection>/<tenant>/<version>` and deleting the old version when updated **after** updating the cache map (so that all new tenants have access to new tenant data). Basically similar to what you proposed.

I'm not 100% sure about what is the right time to delete the old version. Even though every lookup after adding tenant will return new path, still there can be client using old path and deleting can result in client failures. Added TODO about this known constrains and will come back to it if that's causing huge problems.

File: exp/query/lsm_fetcher.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+		c, err := l.cache.Tenant(collection, tenant)
+		if err == nil && version != 0 && c.Version >= version {
+			tenantPath = c.AbsolutePath()
+			download = false
+		}
+		if err == nil && version != 0 && c.Version < version {
+			// oldVersion is used to cleanup previous version of tenant data.
+			oldVersion = c.Version
+		}
+		if err != nil && !errors.Is(err, ErrTenantNotFound) {
```
Comment (reyreaud-l) [inline exp/query/lsm_fetcher.go:97]: NIT: we check both above that err == nil, we can just flip the condition to first check `if err != nil` and then have the other conditions. Makes the cody slightly easier to read and more go-idiomatic by having the err check first.

File: exp/query/cache.go (hunk 4 above)
Comment (reyreaud-l) [reply exp/query/cache.go:101]: Could we mark tenant for "eviction" and when a search lands we ensure that we're not using a tenant marked for eviction ? If we have no alternative version we can "unmark" it for eviction ? 

Just a suggestion and can come as a follow up PR

File: exp/query/cache.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+	return sizeBytes
+}
+
+func (d *DiskCache) evict() error {
+	evictSize := int64(float64(d.maxCap) * evictFraction)
+
+	for d.usage() > (d.maxCap - evictSize) {
+		e := d.evictList.Back()
+		if e == nil {
+			// NOTE: Shouldn't happen.
```
Comment (reyreaud-l) [inline exp/query/cache.go:134]: NIT: Maybe we should log a big error that this shouldn't happen but it is happening ? What could ops do if such a warning appears ?

File: exp/query/lsm_fetcher.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+				return nil, "", fmt.Errorf("failed populate the cache: %w", err)
+			}
+		}
+
+		// remove old version
+		if oldVersion > 0 {
+			// NOTE: Removing old version can be tricky, even though all new lookup returns new version path.
+			// There can be some clients still using old versions. We can leave without removing and delete only if eviction is needed.
+			// But still that doesn't remove above stated problem.
+			if l.remove(ctx, basePath, collection, tenant, oldVersion); err != nil {
```
Comment (reyreaud-l) [inline exp/query/lsm_fetcher.go:124]: Question: could it be feasible to keep track of current search requests and which tenant they're using ? Think an `async.Map` where key = tenant+version and value = atomic of search using it.
 
 On marking for eviction we can forbid new search to that version (or unmark for deletion if the version is still the latest one). Then we trigger eviction on that count hitting 0 and marked for eviction ? With a proper reaper ensuring we don't have runway process holding an eviction happening and we eat the rare failing search. WDYT ? (this can all come as follow up prs, just starting the discussion).

File: exp/query/cache.go (hunk 7 above)
Comment (kavirajk) [reply exp/query/cache.go:134]: good thinking 👍 

I would rather panic here. This means basically our querier cannot store any copy of tenant data, which intern means it cannot serve any query. 

Once we take it stage setup, we can learn from it and see if we can do any workaround. For now I want it to fail immediately to stop making it worse.

File: exp/query/lsm_fetcher.go (hunk 8 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go:124]: Good thinking. One problem with that approach is, we are dealing with "files" not in-memory key-values. So once we return to clients, they can do anything with the local path of that tenant. We have very less control. Basically this part

> keep track of current search requests and which tenant they're using 

is hard in our disk based cache. If we decide to go that path, then we need to track the list of "file handles" for all the files on LSM directory. Worse in my opinion.

> On marking for eviction we can forbid new search to that version

We already do that. Sort of, once the new version of tenant is added, every search after that always returns only the latest copy to any clients (note the `key` in `tenants` map is still `/collection/tenant`). But we return the "local filesystem path" to clients. Which they can store and use it as long as they want. We cannot have any control.

If we want more control, they we have to design cache API that returns contents of each file instead of file path. Would be complex at this point in time I guess.

File: exp/query/lsm_fetcher.go (hunk 6 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go]: Done

File: exp/query/cache.go (hunk 7 above)
Comment (kavirajk) [reply exp/query/cache.go:134]: Done

File: exp/query/lsm_fetcher.go (hunk 5 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go]: Done

File: exp/query/cache.go (hunk 4 above)
Comment (kavirajk) [reply exp/query/cache.go:101]: explained why it's hard to do that [here](https://github.com/weaviate/weaviate/pull/6267#discussion_r1836713824) :) Happy to discuss if i'm missing anything.

File: exp/query/lsm_fetcher.go (hunk 2 above)
Comment (kavirajk) [reply exp/query/lsm_fetcher.go]: Done. singleflight will come in follow up PR

**Review by reyreaud-l — DISMISSED** [review]

File: exp/query/lsm_fetcher.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+	tenantPath := path.Join(
+		basePath,
+		strings.ToLower(collection),
+		strings.ToLower(tenant),
+		versionStr,
+	)
+
+	// src - s3://<collection>/<tenant>/<node>/
+	// dst (local) - <data-path/<collection>/<tenant>
```
Comment (nathanwilk7) [inline exp/query/lsm_fetcher.go:153]:
```suggestion
	// dst (local) - <data-path/<collection>/<tenant>/<version>
```

**Review by nathanwilk7 — APPROVED** [review]
Thanks for addressing the feedback!

**Review by reyreaud-l — APPROVED** [review]
